# Importing Libraries

In [1]:
import pandas as pd
import numpy as np

# Reading data

In [2]:
df=pd.read_csv('Downloads/mushrooms.csv')
df.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

In [4]:
df.describe()

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

In [5]:
df.columns

Index(['class', 'cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat'],
      dtype='object')

# Finding Missing Values

In [6]:
df.isna().sum()

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

So there is no missing values in the dataset

In [18]:
#to sort out categorical values
categorical_features=[feature for feature in df.columns if df[feature].dtypes=='O']
categorical_features
print('The length of categorical features is :',len(categorical_features),'\nThe categorical features are ',categorical_features)

The length of categorical features is : 22 
The categorical features are  ['cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor', 'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color', 'stalk-shape', 'stalk-root', 'stalk-surface-above-ring', 'stalk-surface-below-ring', 'stalk-color-above-ring', 'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number', 'ring-type', 'spore-print-color', 'population', 'habitat']


# Encoding

In [19]:
from sklearn import preprocessing
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder

label_en=LabelEncoder()
df['class']=label_en.fit_transform(df['class'])

In [20]:
df['class'].unique()

array([1, 0], dtype=int64)

In [21]:
df.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0      1         x           s         n       t    p               f   
1      0         x           s         y       t    a               f   
2      0         b           s         w       t    l               f   
3      1         x           y         w       t    p               f   
4      0         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [33]:
# splitting data for training

X = df.drop('class',axis=1)
y = df['class']

# one hot encoding
X=pd.get_dummies(X)

# Scaling the data
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
scaler = StandardScaler()
X = scaler.fit_transform(X)


# creating test train sets
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test =train_test_split(X,y,random_state=42,test_size=0.3) 

In [34]:
from sklearn.linear_model import LogisticRegression
lr=LogisticRegression()
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier()
from sklearn.tree import DecisionTreeClassifier
dt=DecisionTreeClassifier()

In [37]:
classifiers=[('LogisticRegression',lr),('KNeighbors',knn),('DecisionTree',dt)]
for clf_name,clf in classifiers:
    clf.fit(X_train,y_train)
    y_pred=clf.predict(X_test)
    print("Accuracy for: ",clf_name, accuracy_score(y_test,y_pred))
    print("Precision is: ",clf_name, precision_score(y_test,y_pred))
    print("Recall is: ",clf_name, recall_score(y_test,y_pred))
    print("F1 score is: ",clf_name, f1_score(y_test,y_pred))


Accuracy for:  LogisticRegression 1.0
Precision is:  LogisticRegression 1.0
Recall is:  LogisticRegression 1.0
F1 score is:  LogisticRegression 1.0
Accuracy for:  KNeighbors 0.9995898277276456
Precision is:  KNeighbors 1.0
Recall is:  KNeighbors 0.9991532599491956
F1 score is:  KNeighbors 0.9995764506565015
Accuracy for:  DecisionTree 1.0
Precision is:  DecisionTree 1.0
Recall is:  DecisionTree 1.0
F1 score is:  DecisionTree 1.0


# Modelling

In [23]:
# GaussianNB classifier
from sklearn.naive_bayes import GaussianNB
classifierGauss = GaussianNB()
classifierGauss.fit(X_train,y_train)

GaussianNB()

In [24]:
y_pred=classifierGauss.predict(X_test)

In [25]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

print("Accuracy is: ", accuracy_score(y_test,y_pred))
print("Precision is: ", precision_score(y_test,y_pred, average = 'micro'))
print("Recall is: ", recall_score(y_test,y_pred, average = 'micro'))
print("F1 score is: ", f1_score(y_test,y_pred, average = 'micro'))

Accuracy is:  0.9626743232157506
Precision is:  0.9626743232157506
Recall is:  0.9626743232157506
F1 score is:  0.9626743232157506


Very good prediction score is obtained with gaussian classifier.

In [26]:
confusion_matrix(y_pred,y_test)

array([[1167,    1],
       [  90, 1180]], dtype=int64)

In [27]:
# Bernoulli classifier
from sklearn.naive_bayes import BernoulliNB
classifierB = BernoulliNB()
classifierB.fit(X_train,y_train)
y_pred=classifierB.predict(X_test)

In [28]:
print("Accuracy is: ", accuracy_score(y_test,y_pred))
print("Precision is: ", precision_score(y_test,y_pred, average = 'micro'))
print("Recall is: ", recall_score(y_test,y_pred, average = 'micro'))
print("F1 score is: ", f1_score(y_test,y_pred, average = 'micro'))

Accuracy is:  0.9343724364232978
Precision is:  0.9343724364232978
Recall is:  0.9343724364232978
F1 score is:  0.9343724364232978


In [29]:
confusion_matrix(y_pred,y_test)

array([[1241,  144],
       [  16, 1037]], dtype=int64)

Both the classifier performed very well. The Guassian classifier performs slightly better than Bernolli with accuracy of 0.96267 as compared to 0.93437.

 